### PR TITLE
refactor: improve error handling

### PR DESCRIPTION
**Pull request by geek-soft**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/AM-215

### Change description ###

This PR introduces the following improvements:
- JDBI exceptions are wrapped with AM exceptions to keep right abstraction level
- all AM exceptions inherit from base exception to make handling easier

### Checklist ###

- [ ] PR title follows [conventional commits](https://www.conventionalcommits.org) approach and uses one of the following types:
    - build: changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
    - ci: changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
    - docs: documentation only changes
    - feat: a new feature
    - fix: a bug fix
    - perf: a code change that improves performance
    - refactor: a code change that neither fixes a bug nor adds a feature
    - style: changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
    - test: adding missing tests or correcting existing tests
- [ ] commit messages are meaningful and follow good commit message guidelines to ease review process
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

### CODE REVIEW

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/exceptions/PersistenceException.java [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
+ * Exception thrown when operation requiring persistent data store has failed,
+ * usually wrapping exceptions from persistence framework of choice.
+ */
+public class PersistenceException extends AccessManagementException {
     private static final long serialVersionUID = 1L;
 
-    public PersistenceException(Throwable throwable) {
-        super("The transaction has been rolled back. Cause: " + throwable);
+    public PersistenceException(Throwable ex) {
+        super("Operation on persistent store failed. In case operation run in transaction it has been rolled back. "
```
Comment (swalker125) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/exceptions/PersistenceException.java]: this doesn't read right to me. 

"Operation on persistent store failed. If the operation was a transaction, it has been rolled back."

?

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/internal/aspects/ErrorHandlingAspect.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import uk.gov.hmcts.reform.amlib.exceptions.PersistenceException;
+
+@Aspect
+public class ErrorHandlingAspect {
+
+    @Around("execution(public * uk.gov.hmcts.reform.amlib.*Service.*(..))")
+    public Object around(ProceedingJoinPoint joinPoint) throws Throwable {
+        try {
+            return joinPoint.proceed();
+        } catch (JdbiException ex) {
```
Comment (swalker125) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/internal/aspects/ErrorHandlingAspect.java]: is there cases where something other than a Jdbi exception would want us to throw our persistence exception?

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/importer/ResourceDefinitionIntegrationTest.java [hunk 3]
```diff
@@ -16,7 +16,7 @@
 
     @Test
     void shouldNotBeAbleToCreateResourceForServiceThatDoesNotExist() {
-        assertThatExceptionOfType(UnableToExecuteStatementException.class).isThrownBy(() ->
+        assertThatExceptionOfType(PersistenceException.class).isThrownBy(() ->
```
Comment (swalker125) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/importer/ResourceDefinitionIntegrationTest.java]: I believe we have another test similar to this for roles. Would this also need to be updated to PersistenceException?

Comment (geek-soft) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/importer/ResourceDefinitionIntegrationTest.java]: The other test already expects `PersistenceException`.

```
assertThatExceptionOfType(PersistenceException.class)
            .isThrownBy(() -> service.grantDefaultPermission(createDefaultPermissionGrant(READ_PERMISSION)))
            .withMessageContaining("(role_name)=(Role Name) is not present in table \"roles\"");
```

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/internal/aspects/ErrorHandlingAspect.java (hunk 2 above)
Comment (geek-soft) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/internal/aspects/ErrorHandlingAspect.java]: I do not think so. On one hand all low level SQL exceptions are wrapped in one of the JDBI exceptions that all inherit from `JdbiException`. On the other hand `PersistenceException` should be only used when dealing with database or other persistent data store of choice.

**Pull request merged**